### ISSUE TITLE
How To Autorun KrakenX On Linux

### ISSUE THREAD

**Issue by kkobashi** [KsenijaS/krakenx#28]
### TL;DR
Instructions on how to auto run the KrakenX colctl script on bootup. (*)

### Operating System
This is for Linux only

### Purpose
To execute KrakenX script inside a service daemon that gets launched during boot up.

This is very important so you don't have to remember to always run the script after you log in.

With CAM on Windows, everything kicks in when Windows boots up. On Linux however, NZXT provides no driver support or CAM. When Linux boots up, your Kraken AIO will default to a white RGB pump header color and the fans and pump will operate at default settings.

If you want more control than this. KrakenX lets you change this.

What follows are instructions on how to get the latest KrakenX script, install all relevant code, configure your Kraken liquid cooler and administrate the auto run service.

### Get Python and PIP Assets, Setup and Install

sudo apt install python3
sudo apt-get install python3-usb
sudo apt-get install python3-pip

### Get KrakenX Assets, Setup and Install

cd /etc
mkdir colctl
Download Kraken Zip file and copy to /etc/colctl
unzip krakenx-master.zip
cd krakenx-master
python3 -m pip install krakenx (don't sudo or file error with owner cache permission appears)

**Verify Your File Structure**

Before you proceed, make sure your file structure looks similar to this:

/etc
--/colctl
------colctl
--/krakenx
-----color_change.py
-----__init__.py
-----profile.py

These are the minimal folder and files you need to run KrakenX. They must be in this layout or the auto run service will not work.

### Create a System Daemon Service For KrakenX
This involves running a text editor (nano in this case) to type some stuff in. You will save the text file in /etc/systemd/service under the filename krakenx.service. 

The parameter settings below are set for (*):
- SpectrumWave RGB on the pump header
- Fan speed (Celcius, not Fahrenheit):
    Above 20C at 25% level
    Above 30C at 60% level
    Above 40C at 80% level
    Above 50C at 100% level
- Pump speed
    Above 20C at 60% level
    Above 30C at 70% level
    Above 40C at 80% level
    Above 50C at 100% level

Adjust the above to your liking.

cd /etc/systemd/service
sudo nano krakenx.service

[Unit]
Description=Kraken AIO startup service

[Service]
Type=oneshot
User=root
WorkingDirectory=/etc/colctl
ExecStart=/etc/colctl/colctl --mode SpectrumWave --fan_speed "(20,25),(30,60),(40,80),(50,100)" --pump_speed "(20,60),(30,70),(40,80),(50,100)"

[Install]
WantedBy=multi-user.target

### Set Permissions and Ownership

sudo chmod 644 krakenx.service (must be 644 or syslog will show errors)
sudo chown root:root krakenx.service
chmod 755 /etc/colctl/colctl
chown root:root /etc/colctl/colctl

### Using KrakenX Service

Ready? Set. Go!

This first command starts the service and puts it in running state.
The second command enables it so that on next boot, it will run automatically.

sudo systemctl start krakenx.service
sudo systemctl enable krakenx.service

You should now see your Kraken RGB, fans, and pump run at your specified settings.

### Other Things You May Need To Know

These are things that you may need to use later to administrate the service. If you make changes to any of the parameter settings above, you need to stop the service, update systemd, and restart the service again.

I leave them here for reference.

**_To start the service_**
systemctl start krakenx.service
systemctl enable krakenx.service
	
**_To stop the service_**
systemctl stop krakenx.service
systemctl disable krakenx.service

**_To restart the service_**
systemctl restart kraken.service
	
**_To reload the service_**
systemctl reload krakenx.service
	
**_To restart and reload the service_**
systemctl reload-or-restart krakenx.service
	
**_To remove the service_**
systemctl stop krakenx.service
systemctl disable krakenx.service
rm /etc/systemd/system/krakenx.service
rm /etc/systemd/system/krakenx.service (don't forget symlinks)

**_To update systemd_**
systemctl daemon-reload
systemctl reset-failed
	
**_Status, targets and properties_**
systemctl get-default or ls -al /lib/systemd/system/default.target
ls -al /lib/systemd/system/runlevel*
systemctl list-unit-files
systemctl list-units --type=target
systemctl is-enabled krakenx.service
systemctl is-failed krakenx.service
systemctl is-active krakenx.service
systemctl show krakenx.service

Hope this helps someone. Feel free to add to the project README.

Note: This was edited several times to arrive at this solution. I do not recommend using cron or setting colctl in shell startup scripts. I had problems getting them to work. This is much cleaner, set and forget solution.

**_(*) Use at your own risk. I will not be held responsible for anything that may go wrong. This was developed and tested on Ubuntu 18.0.4 desktop._**

**Comment by jonasmalacofilho** [comment]
Thanks!

This should be included in the documentation.

---

@jneumaier where do we want to put the growing documentation: `README` only, `/docs` or Wiki?

**Comment by jneumaier** [comment]
@kkobashi Thank you!

Good idea, I guess everyone needs this somehow. I might add a Windows version using the task scheduler.

@jonasmalacofilho I would prefer to split the extended documentation in some extra place and link it from `README.` Maybe a wiki would be a clear favorite for a bigger number of documents. In our case I would prefer a `/docs` directory but I am fine with both.

**Comment by kkobashi** [comment]
I've updated this to use a system daemon instead. There's too much to consider regarding interactive, non-interactive, non-login, login cases and the order the scripts are executed. Cron upon reboot, even every 1 minute didn't trigger colctl. I think it may have to do with udev rules, USB, and permissions.

**Comment by jneumaier** [comment]
@kkobashi I integrated you documentation. Please feel free to reopen this issue or create a pull request any time.

**Comment by kkobashi** [comment]
@jneumaier When you copy and pasted, the formatting got messed up. I tried fixing but I'm prevented from doing so. Please fix so its readable in the same format as I have here. In particular, the commands are all run together on one line. They need CR/LF probably as they were probably stripped by your Windows editor.

**Comment by jneumaier** [comment]
@kkobashi Git manages line endings even if Windows users mess them up :stuck_out_tongue: They were all fine, but I thought it was valid markdown syntax already which handles breaks automaticallly. Sorry, should have checked this more thorough.

jonasmalacofilho had some suggestions for improvements he left in commit comments, see https://github.com/KsenijaS/krakenx/commit/6bf4f925a187e4b51df29b81cd5da621a37d9b29
You can always create a pull request from your own fork:
https://help.github.com/articles/creating-a-pull-request-from-a-fork/

**Comment by jneumaier** [comment]
We should reference the default installation procedure in [doc/linux-autorun.md](https://github.com/KsenijaS/krakenx/blob/master/doc/linux-autorun.md). Otherwise this will be hard to maintain.

Can we not just install as described in README.md and do
`ExecStart=colctl`
Don't we have a PATH resolval here? What is the reason for this custom installation procedure?

**Comment by jonasmalacofilho** [comment]
There are a few quirks – e.g. systemd units don't see the same environment (variables) as you do – but it can usually be managed without installing things twice or requiring them to be in `/etc`.

Though I haven't tested this with krakenx,[1] and @kkobashi also mentions a possible (and likely) dependency issue with udev.

[1] I'm traveling for work, I'm not sure I'll be able to test this before I get back.

**Comment by jonasmalacofilho** [comment]
Closing, as there is nothing to be done _directly_.

I've been testing a [simpler](https://github.com/jonasmalacofilho/dotfiles/tree/master/krakenx) setup.  Eventually, I'll _propose_ a complete rewrite of the instructions based on this simpler approach, as the custom installation (and inside `/etc`, of all places) really does not sit well with me.

Note: systemd 239.3030-1 introduced [some](https://github.com/jonasmalacofilho/dotfiles/commit/e721bfeb36590f2b11bc2bea7d537e607e2c634e#diff-e97b0c0348098daf807e74b3ad362bb8) issue related to the initialization of the device (not dbus).  I already researched how to solve that [properly](https://lists.fedoraproject.org/pipermail/devel/2012-January/160917.html), but haven't yet had the time to test it.
